### PR TITLE
NodeJS: fix duplicates

### LINKED ISSUE #628: NodeJS: fix duplicates
### Description

NodeJS output.txt file has a lot of duplicates that need to be removed.
One way to solve this issue is to introduce a Dictionary to keep track of Articles and avoid duplicates.

## Steps to recreate

Run tests in duckpan:
`DDG_TEST_FATHEAD=node_js prove --quiet t/validate_fathead.t 1>/dev/null`

https://gist.github.com/bfmags/86f81caf513c0078b1d3be3918c22526

Refer to documentation : 
https://docs.duckduckhack.com/programming-mission/creating-effective-fatheads.html

## People to notify

@MariagraziaAlastra @tagawa @moollaza 

## Get Started
- [ ] 1) Claim this issue by commenting below
- [ ] 2) Review our [Contributing Guide](https://github.com/duckduckgo/zeroclickinfo-fathead/blob/master/CONTRIBUTING.md)
- [ ] 3) [Set up your development environment](https://docs.duckduckhack.com/welcome/setup-dev-environment.html), and fork this repository
- [ ] 4) Create a Pull Request

## Resources
- Join [DuckDuckHack Slack](https://quackslack.herokuapp.com/) to ask questions
- Join the [DuckDuckHack Forum](https://forum.duckduckhack.com/) to discuss project planning and Instant Answer metrics
- Read the [DuckDuckHack Documentation](https://docs.duckduckhack.com/) for technical help

---

Instant Answer Page: https://duck.co/ia/view/node_js

### PR DESCRIPTION
**Pull request by ahetawal-p**
## Description of new Instant Answer, or changes
Fixing duplicates in output.txt for nodejs documentation IA.

## Related Issues and Discussions
Fixes #628

## People to notify
@MariagraziaAlastra @tagawa @moollaza @bfmags 

## Testing & Review
To be completed by Language Leader (or DDG Staff)

**Pull Request**
- [x] Title follows correct format (Specifies Instant Answer + Purpose)
- [x] Description contains a valid Instant Answer Page Link (e.g. https://duck.co/ia/view/my_ia)

**Instant Answer Page**
- [ ] Instant Answer page is correctly filled out and contains:
    - One topic for the Search Space Language (Java, Python, Scala, Ruby, etc.)
    - One topic from: Reference, Help, Libraries, Tools
        - Documentation Fatheads are considered "Reference"
    - Description, Data source, and 2+ example queries
    - Perl Module (e.g. "DDG::Fathead::PerlDoc" -- we only need a name, not an actual file)
    - Source Name (for "More at <source_name>" link)
    - Source Domain (must contain http:// or https:// -- can be the same as Data Source)
    - Source Info (used as Subtitle for each Article -- usually matches the IA Name)
    - 'Skip Abstract' is checked off
    - Source ID (ping @moollaza to assign one, once Fathead is ready for Beta deploy)
- [ ] Fathead Tests are passing (run `$ duckpan test <fathead_id>`)
    - Tester should report any failures

**Code**
- [x] Uniformly indented, well commented
- [x] Fetch.sh and Parse.sh run without errors
- [x] Output contains no blank lines, or multi-line entries

**Pull Request Review Guidelines**: https://docs.duckduckhack.com/programming-mission/pr-review.html

---

Instant Answer Page: https://duck.co/ia/view/node_js

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by bfmags** [comment]
@ahetawal-p just noticed some of the duplicates should be disambiguations (articles with same name pointing to different pages, like e.g. [` Buffer.from	A`](https://github.com/duckduckgo/zeroclickinfo-fathead/pull/632/files#diff-f2695be7819d8d1b3a17c17f6dc38f3eL279) 

If someone searches for `Buffer.from` we need to show all relevant articles.

* First we need to differentiate all the dups, in case of "Buffer.from" they have different bookmarks #all_buffer_from_buffer,#all_buffer_from_arraybuffer_byteoffset_length,etc so we can use those

e.g. `Buffer.from(buffer)    A`, `Buffer.from(arraybuffer_byteoffset_length)    A`

* Create [Disambiguation](https://docs.duckduckhack.com/resources/fathead-overview.html) entry, consisting of a list where each item is an article title, followed by a one sentence description ending in a period. The items must be separated by \n*.

e.g. `length	D								*[[AudioBuffer.length]] A float.\n*[[CSSValueList.length]] The length read-only property of the CSSValueList interface represents the number of CSSValue s in the list.\n*[[DataTransferItemList.length]] The read-only DataTransferItemList.\n*[[History.length]] The History.\n*[[HTMLFormElement.length]] The HTMLFormElement.\n*[[NodeList.length]] length returns the number of items in a NodeList.`

**Comment by ahetawal-p** [comment]
Thanks @bfmags 
Just so that I understand it correctly :
If a user searches for Buffer.from

The entries in output.txt will consists of - 
Buffer.from(buffer) A
Buffer.from(arraybuffer_byteoffset_length) A
..
..
..
Buffer.from A  "\t" \"\t ...."point to above disambiguation entries""\t"
etc...

So only the Buffer.from will have entries in its disambiguation column to point to the above dups .?
Is 
And also users can still search through keys of something like this "Buffer.from(buffer)"?
What can be an easy way to find these disambiguations, just look at the diff ?

**Comment by bfmags** [comment]
Hey, yeah :)

`Buffer.from D "\t" "\t ...."point to above disambiguation entries""\t"
etc...`

**Comment by ahetawal-p** [comment]
Hi @bfmags,
I have made the changes for adding the disambiguation for similar entries. Do you mind taking a look.

Also I am trying to test the new 'D' entries but I dont seem to get any results back. Could you tell me if the format for these entries is correct or not ?
Ex:  https://beta.duckduckgo.com/?q=node+Buffer 
https://github.com/duckduckgo/zeroclickinfo-fathead/pull/632/files#diff-f2695be7819d8d1b3a17c17f6dc38f3eR1

This should have given me a D entry in search but I am not able to see it?

Thank you,

**Comment by ahetawal-p** [comment]
Seems to be coming up fine now for disambiguations:

https://beta.duckduckgo.com/?q=node+Buffer+from

Let me know if this looks okay @bfmags

File: lib/fathead/node_js/output.txt [hunk 1]
```diff
@@ -1078,24 +1029,15 @@ net Server	R	net.Server										http://nodejs.org/api/all.html#all_net server
...
+net.Server.close	D								*[[net.Server.close(all_net.server_close)]] Emitted when the server closes. Note that if connections exist, this event is not emitted until all connections are ended.\n*[[net.Server.close(all_net.server_close_callback)]] Stops the server from accepting new connections and keeps existing connections. This function is asynchronous, the server is finally closed when all connections are ended and the server emits a &apos;close&apos; event. The optional callback will be called once the &apos;close&apos; event occurs. Unlike that event, it will be called with an Error as its only argument if the server was not open when it was closed.			
 net.Server close	R	net.Server.close										http://nodejs.org/api/all.html#all_net.server_close_callback
 net Server close	R	net.Server close										http://nodejs.org/api/all.html#all_net server_close_callback
```
Comment (bfmags) [inline lib/fathead/node_js/output.txt]: Hi @ahetawal-p sorry for the delay. Great progress on solving the dups, but there's still some around ([check gist](https://gist.github.com/bfmags/8157f03d38f1ec5d556ead78361cc5a1))
eg. `net.Server close	R`
Thanks!

Comment (ahetawal-p) [inline lib/fathead/node_js/output.txt]: Hey @bfmags Sorry I didnt get you ? I ran the validation script on my side and I dont see any diffs.
Based on your gist if I compare the line numbers on the new output.txt generated after my changes I dont see the dup line numbers matching. Can you try running the validation script on the fresh output.txt may be ???

Comment (bfmags) [inline lib/fathead/node_js/output.txt]: @ahetawal-p you're right!

**Pull request merged**